runtime/gc_vlhgc/AllocationContextBalanced.cpp

```diff
@@ -1103,6 +1103,56 @@ MM_AllocationContextBalanced::setNumaAffinityForThread(MM_EnvironmentBase *env)
 }
 
 #if defined(J9VM_GC_SPARSE_HEAP_ALLOCATION)
+bool
+MM_AllocationContextBalanced::allocateFromSharedArrayReservedRegion(MM_EnvironmentBase *env, uintptr_t fraction)
+{
+	const uintptr_t regionSize = _heapRegionManager->getRegionSize();
+	Assert_MM_true((regionSize > fraction) && (0 != fraction));
+
+	bool shouldAllocateNewSharedRegion = false;
+
+	lockCommon();
+	uintptr_t oldBytesUsed = _sharedArrayReservedRegionsBytesUsed;
+	uintptr_t wholeRegionsBefore = oldBytesUsed / regionSize;
+	_sharedArrayReservedRegionsBytesUsed += fraction;
+	uintptr_t wholeRegionsAfter = _sharedArrayReservedRegionsBytesUsed / regionSize;
+
+	if ((0 == oldBytesUsed) || (wholeRegionsBefore < wholeRegionsAfter)) {
+		shouldAllocateNewSharedRegion = true;
+	}
+	unlockCommon();
+
+	return shouldAllocateNewSharedRegion;
+}
+
+bool
+MM_AllocationContextBalanced::recycleToSharedArrayReservedRegion(MM_EnvironmentBase *env, uintptr_t fraction)
+{
+	const uintptr_t regionSize = _heapRegionManager->getRegionSize();
+	Assert_MM_true((regionSize > fraction) && (0 != fraction) && (_sharedArrayReservedRegionsBytesUsed >= fraction));
```

Comment:
break it into 2 asserts. `_sharedArrayReservedRegionsBytesUsed >= fraction `should be a separate one